### PR TITLE
Fix shared pointer factory create 

### PR DESCRIPTION
**Pull request by dschwen**
## Reason
The Apollo code uses `std::enable_shared_from_this`, and now throws an exception since https://github.com/idaholab/moose/pull/26945/files#diff-42b65125f9d80b395d9ba504930bc4e95069965a6a7014069ee5524a746edb26L91-L101

## Design
Create a shared pointer properly.

## Impact
Fix Apollo.

### CODE REVIEW

**Review by loganharbour — APPROVED** [review]

**Comment by loganharbour** [comment]
> properly

😞

**Comment by dschwen** [comment]
I'll build Apollo against this patch to confirm it fixes the issue.

**Comment by dschwen** [comment]
App failures are unrelated, module failure is an unrelated timeout

**Comment by dschwen** [comment]
To paraphrase a certain Mountain Dew flavor: @loganharbour did nothing wrong

**Comment by dschwen** [comment]
Actually he did. The signature of the shared pointer constructor from a raw pointer is templated on the raw pointer type. That raw pointer type must have the `std::enable_shared_from_this` mix-in. The current design is not feasible because the raw pointer is always of a high-up base class type. I'm adding a PR to change the Registry item build methods back to shared pointers.

File: framework/include/base/Factory.h [hunk 1]
```diff
@@ -251,13 +251,7 @@ Factory::createUnique(const std::string & obj_name,
...
                       const THREAD_ID tid)
 {
-  auto object = createUnique(obj_name, name, parameters, tid, false);
-  if (!dynamic_cast<T *>(object.get()))
-    mooseError("We expected to create an object of type '" + demangle(typeid(T).name()) +
-               "'.\nInstead we received a parameters object for type '" + obj_name +
-               "'.\nDid you call the wrong \"add\" method in your Action?");
-
-  return std::unique_ptr<T>(static_cast<T *>(object.release()));
+  return std::unique_ptr<T>(create<T>(obj_name, name, parameters, tid).get());
```
Comment (loganharbour) [inline framework/include/base/Factory.h:254]: this doesn't release the shared ptr

Comment (loganharbour) [reply framework/include/base/Factory.h:254]: oh, I guess it does when it goes out of scope... ugh this is so dirty

Comment (loganharbour) [reply framework/include/base/Factory.h:254]: wait... no. this doesn't make sense. you'll get a ~shared_ptr when this leaves scope and the end up with a unique_ptr that points to... a destructed object

**Comment by dschwen** [comment]
`nearest_point_layered_average` oh no, that nightmare of an object again...

File: framework/include/base/Factory.h (hunk 1 above)
Comment (dschwen) [reply framework/include/base/Factory.h:254]: uhm... yeah, dang

**Comment by dschwen** [comment]
I'll give putting enable shared from this into MooseObject (or whatever the base class is these days) a try today

**Comment by dschwen** [comment]
Looks like there is a little rebase screw-up. I'll let these tests run their course and fix it later.

**Comment by dschwen** [comment]
Anybody not in :ecuador: who could comment on this simpler fix?

**Review by loganharbour — APPROVED** [review]

**Pull request merged**